docs/src/rules/no-useless-assignment.md

```diff
@@ -0,0 +1,145 @@
+---
+title: no-useless-assignment
+rule_type: suggestion
+related_rules:
+- no-unused-vars
+further_reading:
+- https://en.wikipedia.org/wiki/Dead_store
+- https://www.npmjs.com/package/eslint-plugin-no-useless-assign
+- https://sonarsource.atlassian.net/browse/RSPEC-1854
+- https://rules.sonarsource.com/javascript/RSPEC-1854/
+- https://rules.sonarsource.com/typescript/RSPEC-1854/
+- https://cwe.mitre.org/data/definitions/563.html
+- https://wiki.sei.cmu.edu/confluence/display/c/MSC13-C.+Detect+and+remove+unused+values
+- https://wiki.sei.cmu.edu/confluence/display/java/MSC56-J.+Detect+and+remove+superfluous+code+and+values
+---
+
+
+[Wikipedia describes a "dead store"](https://en.wikipedia.org/wiki/Dead_store) as follows:
+
+> In computer programming, a local variable that is assigned a value but is not read by any subsequent instruction is referred to as a **dead store**.
+
+"Dead stores" waste processing and memory, so it is better to remove unnecessary assignments to variables.
+
+Also, if the author intended the variable to be used, there is likely a mistake around the dead store.
+
+## Rule Details
+
+This rule aims to report variable assignments when the value is not used.
+
+Examples of **incorrect** code for this rule:
+
+::: incorrect
+
+```js
+/* eslint no-useless-assignment: "error" */
+
+function foo() {
+    let v = 'used';
+    doSomething(v);
+    v = 'unused';
+}
+
+function foo() {
+    let v = 'used';
+    if (condition) {
+        v = 'unused';
+        return
+    }
+    doSomething(v);
+}
+
+function foo() {
+    let v = 'used';
+    if (condition) {
+        doSomething(v);
+    } else {
+        v = 'unused';
+    }
+}
+
+function foo() {
+    let v = 'unused';
+    if (condition) {
+        v = 'used';
+        doSomething(v);
+        return
+    }
+}
+
+function foo() {
+    let v = 'used';
+    if (condition) {
+        let v = 'used';
+        console.log(v);
+        v = 'unused';
+    }
+    console.log(v);
+}
+```
```

Comment:
As it is now, the code in this example cannot be parsed in the Playground with the default `{ "sourceType": "module" }` because of the duplicated function name `foo` ([see here](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZWxlc3MtYXNzaWdubWVudDogXCJlcnJvclwiICovXG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgICBsZXQgdiA9ICd1c2VkJztcbiAgICBkb1NvbWV0aGluZyh2KTtcbiAgICB2ID0gJ3VudXNlZCc7XG59XG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgICBsZXQgdiA9ICd1c2VkJztcbiAgICBpZiAoY29uZGl0aW9uKSB7XG4gICAgICAgIHYgPSAndW51c2VkJztcbiAgICAgICAgcmV0dXJuXG4gICAgfVxuICAgIGRvU29tZXRoaW5nKHYpO1xufVxuXG5mdW5jdGlvbiBmb28oKSB7XG4gICAgbGV0IHYgPSAndXNlZCc7XG4gICAgaWYgKGNvbmRpdGlvbikge1xuICAgICAgICBkb1NvbWV0aGluZyh2KTtcbiAgICB9IGVsc2Uge1xuICAgICAgICB2ID0gJ3VudXNlZCc7XG4gICAgfVxufVxuXG5mdW5jdGlvbiBmb28oKSB7XG4gICAgbGV0IHYgPSAndW51c2VkJztcbiAgICBpZiAoY29uZGl0aW9uKSB7XG4gICAgICAgIHYgPSAndXNlZCc7XG4gICAgICAgIGRvU29tZXRoaW5nKHYpO1xuICAgICAgICByZXR1cm5cbiAgICB9XG59XG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgICBsZXQgdiA9ICd1c2VkJztcbiAgICBpZiAoY29uZGl0aW9uKSB7XG4gICAgICAgIGxldCB2ID0gJ3VzZWQnO1xuICAgICAgICBjb25zb2xlLmxvZyh2KTtcbiAgICAgICAgdiA9ICd1bnVzZWQnO1xuICAgIH1cbiAgICBjb25zb2xlLmxvZyh2KTtcbn0iLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)). This can be easily fixed by giving the functions unique names. That's what we did in https://github.com/eslint/eslint/pull/17633/files in other places. Since this is a new rule, it would be good to have the examples working in the Playground right from the start.